src/care/host_device_map.h

```diff
@@ -0,0 +1,189 @@
+//////////////////////////////////////////////////////////////////////////////////////
+// Copyright 2022 Lawrence Livermore National Security, LLC and other CARE developers.
+// See the top-level LICENSE file for details.
+//
+// SPDX-License-Identifier: BSD-3-Clause
+//////////////////////////////////////////////////////////////////////////////////////
+#ifndef _CARE_DEVICE_UNORDERED_MAP_H
+#define _CARE_DEVICE_UNORDERED_MAP_H
+#include "care/config.h"
+#include "care/atomic.h"
+
+#include <unordered_map>
+#include <care/KeyValueSorter.h>
+
+namespace care {
+
+   ///
+   /// @author Peter Robinson
+   ///
+   /// @class host_device_map is a rudimentary associative map. On the host,
+   /// it is backed by a std::unordered_map, and on the device it uses a
+   /// care::KeyValueSorter.
+   ///
+   /// Restrictions when compared to std::map:
+   ///   1. Total capacity must be declared up front.
+   ///   2. a call to sort() must be performed before any lookups are done.
+   ///   3. Insertions can only be done via an emplace(Key,Value).
+   ///   4. Lookups can only be done via an at(Key)
+   ///
+   template <typename key_type,
+             typename mapped_type,
+             typename Exec>
+   class host_device_map
+   {
+      public:
+         host_device_map(size_t max_entries);
+         CARE_HOST_DEVICE inline void emplace(key_type key, mapped_type val) const;
+         CARE_HOST_DEVICE inline mapped_type at(key_type key) const;
+         void sort();
+         void free();
+   };
+
+// ********************************************************************************
+// RAJA::seq_exec specialization.
+// ********************************************************************************
+   template <typename key_type, typename mapped_type>
+   class host_device_map< key_type, mapped_type, RAJA::seq_exec> {
+      public:
+         // constructor
+         host_device_map(size_t max_entries) : m_map()  {
+            m_map = new std::unordered_map<key_type, mapped_type>{};
+         }
+        // emplace a key value pair
+        inline void emplace(key_type key, mapped_type val) const {
+           (*m_map)[key] = val;
```

Comment:
Why not call `m_map->emplace(key, val)`?

Reply:
I guess maybe this fails if a key shows up more than once?

Reply:
fixed